Fix tsnr afterburner crash when no cframes on a night

akremin opened the pull request:
### Overview
This fixes #1345, which is a long-standing bug in the tsnr afterburner (`desi_tsnr_afterburner`) that causes it to crash if no cframes are available on a night. This is generally not an issue for reprocessing since we don't process or care about nights without valid science data, but for daily it matters since the crash prevents the exposures and tiles from being listed in the output files and therefore the survey operations don't have a way to propagate the 0 efftime for those exposures.

This fix is imperfect, as it only works if you supply an existing `exposures-${SPECPROD}.fits` file. This solves all existing issues, however, since we can generate that file with any other night in the production that has cframes prior to running on a problematic night. 

### Testing
I tested on all nights linked to that issue: 20210710 , 20210922, 20211213, 20211226

The output files can be found here: `/global/cfs/cdirs/desi/users/kremin/PRs/tsnr_nocframe/`
And include: `tiles-daily.csv `, `exposures-daily.csv`, and `exposures-daily.fits`.

#### Example: Night 20211213

Running out of the `daily` branch
```
kremin@cori09:/global/cfs/cdirs/desi/users/kremin/PRs/tsnr_nocframe> desi_tsnr_afterburner -o /global/cfs/cdirs/desi/users/kremin/PRs/tsnr_nocframe/exposures-daily.fits --prod daily --tile-completeness /global/cfs/cdirs/desi/users/kremin/PRs/tsnr_nocframe/tiles-daily.csv --aux /global/cfs/cdirs/desi/survey/observations/SV1/sv1-tiles.fits --gfa-proc-dir /global/cfs/cdirs/desi/survey/GFA/ --compute-skymags --update --nproc 8 --nights 20211213 --add-badexp
```

Yields: 
```
Traceback (most recent call last):
  File "/global/common/software/desi/cori/desiconda/20211217-2.0.0/code/desispec/main/bin/desi_tsnr_afterburner", line 1211, in <module>
    sys.exit(main())
  File "/global/common/software/desi/cori/desiconda/20211217-2.0.0/code/desispec/main/bin/desi_tsnr_afterburner", line 1045, in main
    tsnr2_expid_table = vstack(tsnr2_expid)
  File "/global/common/software/desi/cori/desiconda/20211217-2.0.0/conda/lib/python3.9/site-packages/astropy/table/operations.py", line 646, in vstack
    tables = _get_list_of_tables(tables)  # validates input
  File "/global/common/software/desi/cori/desiconda/20211217-2.0.0/conda/lib/python3.9/site-packages/astropy/table/operations.py", line 53, in _get_list_of_tables
    raise ValueError('no values provided to stack.')
ValueError: no values provided to stack.
```

And we see the tables don't include the night in question:
```
kremin@cori09:/global/cfs/cdirs/desi/users/kremin/PRs/tsnr_nocframe> grep 20211213 exposures-daily.csv
```

Running out of the branch for this PR, the same command that crashed above now succeeds:
```
kremin@cori09:/global/cfs/cdirs/desi/users/kremin/PRs/tsnr_nocframe> desi_tsnr_afterburner -o /global/cfs/cdirs/desi/users/kremin/PRs/tsnr_nocframe/exposures-daily.fits --prod daily --tile-completeness /global/cfs/cdirs/desi/users/kremin/PRs/tsnr_nocframe/tiles-daily.csv --aux /global/cfs/cdirs/desi/survey/observations/SV1/sv1-tiles.fits --gfa-proc-dir /global/cfs/cdirs/desi/survey/GFA/ --compute-skymags --update --nproc 8 --nights 20211213 --add-badexp
```

Which successfully runs, and we see the rows are in the output tables. They are mostly zeros, as expected because they are 0 efftime exposures that have been flagged as bad in the pipeline:
```
kremin@cori09:/global/cfs/cdirs/desi/users/kremin/PRs/tsnr_nocframe> grep 20211213 exposures-daily.csv
20211213,113678,21281,336.495,21.243,59562.087959013,main,bright,bright,mainbright,718.3,0.0,180.0,bright,0.85,0.0,0.047,0.0,0.1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.084,1.366,0.447,0.338,0.159,0.061,0.064,0.066,1.124,17.207,17.904,17.206,16.856,0.0,0.0,0.0,0.0
20211213,113679,42991,18.994,70.638,59562.101930428,main,backup,backup,mainbackup,607.6,0.0,60.0,backup,0.85,0.0,0.532,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.065,1.501,0.382,0.295,0.141,0.039,0.042,0.043,1.288,17.156,17.858,17.153,16.806,0.0,0.0,0.0,0.0
20211213,113680,42992,24.153,73.579,59562.110868342,main,backup,backup,mainbackup,606.6,0.0,60.0,backup,0.85,0.0,0.448,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.059,1.484,0.396,0.305,0.146,0.034,0.036,0.037,1.347,17.27,17.978,17.271,16.91,0.0,0.0,0.0,0.0
20211213,113681,42959,29.175,70.617,59562.119814932,main,backup,backup,mainbackup,607.2,0.0,60.0,backup,0.85,0.0,0.777,0.0,0.2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.478,1.568,0.364,0.284,0.137,0.293,0.313,0.327,1.293,18.198,18.742,18.195,17.647,0.0,0.0,0.0,0.0
20211213,113682,42997,31.459,76.332,59562.128779753,main,backup,backup,mainbackup,606.6,0.0,60.0,backup,0.85,0.0,0.724,0.0,0.4,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.684,1.695,0.324,0.257,0.125,0.355,0.387,0.41,1.408,18.662,19.093,18.66,17.974,0.0,0.0,0.0,0.0
20211213,113683,42958,35.967,73.209,59562.137704338,main,backup,backup,mainbackup,608.9,0.0,60.0,backup,0.85,0.0,0.732,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.03,1.546,0.368,0.286,0.137,0.019,0.019,0.02,1.343,16.962,17.699,16.959,16.582,0.0,0.0,0.0,0.0
```

marcelo-alvarez commented on bin/desi_tsnr_afterburner:
```diff
@@ -291,192 +291,205 @@ def update_table(table1,table2,keys) :
+        cam_summary = Table(names=preexisting_tsnr2_frame_table.colnames,
+                            dtype=preexisting_tsnr2_frame_table.dtype)
+        cam_summary.meta['EXTNAME'] = 'FRAMES'
+        exp_summary = Table(names=preexisting_tsnr2_expid_table.colnames,
```
When `add_badexp==True`, `exp_summary` is combined with `preexisting_tsnr2_expid_table` using the `astropy.Table.vstack` method when `update_table` is called, currently on line 545 (532) in this branch (main) below:
```
 exp_summary = update_table(preexisting_tsnr2_expid_table,exp_summary,["EXPID"])
```

In the main branch, column entries that are in `preexisting_tsnr2_expid_table` but NOT in `exp_summary`, such as `SKY_MAG_G_SPEC`, `SKY_MAG_R_SPEC`, and `SKY_MAG_Z_SPEC` (because they were not present in `entry`), are masked by the `astropy.table.vstack` method, and consequently "nan" is written in the output exposures csv file for bad exposures.

In this branch, however, whenever there is no good exposure in a night to begin with, `len(summary_rows)==0`, and this new block of code is executed instead, so *by definition* all the columns in `preexisting_tsnr2_expid_table` are present in `exp_summary`. Consequently, when `update_table` is called below, values in columns such as `SKY_MAG_G_SPEC`, `SKY_MAG_R_SPEC`, and `SKY_MAG_Z_SPEC` are zero, instead of "nan". 

This may or may not be an issue that requires further changes but, if it is, the information above may be helpful in resolving it.

marcelo-alvarez commented:
@akremin thanks for this, I've tested it and it does indeed prevent the crash from occurring. 

The only possible issue I noticed is that certain values in rows corresponding to bad exposures in the output exposures csv file (and possibly the fits file, though I did not check explicitly) are different between this branch and the main branch. For example, in the main branch bad exposures have "nan" for columns `SKY_MAG_G_SPEC`, `SKY_MAG_R_SPEC`, and `SKY_MAG_Z_SPEC`, while in this branch they are zero, instead. I gave an explanation for why this happens with an inline comment.

I'm not sure if it matters, since there may be no downstream code that uses the affected columns for exposures that have been marked bad. 

Otherwise, it looks good to me and I'm happy to merge, assuming someone else more familiar with how these files are used (such as @akremin @sbailey or @julienguy) can confirm that there are no negative downstream effects from the change in output in this branch from main.

sbailey commented:
Good catch @marcelo-alvarez on the NaN vs. 0.  Summarizing a conversation with @akremin: either is probably ok, but this PR is in a state where nights with a mix of bad and good exposures get NaN, while nights with entirely bad exposures get 0, and that mismatch is distasteful.  Given that 0 is a valid value for a magnitude and we're currently using NaN, let's update the code to always set SKY_MAG_[GRZ]_SPEC=NaN in the same place where it is also setting the other values for the bad exposures around line 433 (side note: if the sky ever really was magnitude 0, we've got other problems...).

Also, given that main now uses desimodel/0.18.0 for testing with the fix for scipy/1.10.0, let's remove the scipy==1.7.3 version pin from this branch (tests here will fail again, but ok).

sbailey commented:
Also note that there is a potentially non-trivial merge conflict in desi_tsnr_afterburner to resolve.  This might be a side effect of fixing the previous 2-space indentation (!), but it's unclear to me why that would cause a merge conflict or if there is something more subtle.

akremin commented:
Thank you for the careful review, @marcelo-alvarez . As Stephen suggested I have now explicitly set the sky magnitudes to np.nan in the code. I have also added new logic to determine what cameras were available for a given exposure, which is used to populate the `FRAMES` hdu. That was missing in my previous commit. Finally, I removed the update to requirements.txt. The tests will fail in this PR but are fixed in both `main` and `daily`.

@marcelo-alvarez could you please check to make sure my fix solves your issue and that the new FRAMES entries make sense? I'm happy to merge when you're happy to merge. There will be conflicts to resolve, but hopefully nothing that can't be overcome.

akremin commented:
The merge conflict was git confusion over me correcting the two-space indentation. The lines were identical. I will merge with the conflict resolved and then immediately test main to make sure I didn't break anything.
